### PR TITLE
Added visualize_classification_groups.py

### PR DESCRIPTION
**Pull request by jacksonjacobs1**
The code corresponding to the soon-to-be blog post:
"Generate a Powerpoint for Visualizing Classifier Output"

### CODE REVIEW

File: visualize_classification_groups.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+				im = Image.open(img_buf)
+				addimagetoslide(slide, img_buf, top=Inches(coord[0]*1.2+1),left=Inches(coord[1]*1.2 + 0.25), width=Inches(im.height/im.width), height=Inches(1))
+				im.close()
+	ppt.save(pptx_save_path)
+
+if __name__ == "__main__":
+	parser = argparse.ArgumentParser()
+	parser.add_argument('--h5path', type=str, help='The path to an h5 file that contains "patch", "ground_truth_label", and "fname" datasets.')	
+	parser.add_argument('--model_checkpoint', '-c', type=str, default=None, help='The path to a model checkpoint for the torch.load() method.')
+	parser.add_argument('--ppt_save', type=str, help='The full path to save the generated powerpoint.')
```
Comment (choosehappy) [inline visualize_classification_groups.py]: you should think about what parameters are optional and what parameters are required, those which are required are typically positional , with flags being used mostly for modifying the overall behavior.

File: visualize_classification_groups.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+def addimagetoslide(slide,image_stream,left,top, height, width, resize = 1.0, comment = ''):
+	# res = cv2.resize(img , None, fx=resize,fy=resize ,interpolation=cv2.INTER_CUBIC) #since the images are going to be small, we can resize them to prevent the final pptx file from being large for no reason
+	# image_stream = BytesIO()
+	# Image.fromarray(res).save(image_stream,format="PNG")
+	slide.shapes.add_picture(image_stream, left, top ,height,width)
+	txBox = slide.shapes.add_textbox(left, Inches(1), width, height)
+	tf = txBox.text_frame
+	tf.text = comment
+
+def cartesian_product(*arrays):
```
Comment (choosehappy) [inline visualize_classification_groups.py]: seems like this can be removed and use meshgrid instead?
https://numpy.org/doc/stable/reference/generated/numpy.meshgrid.html

File: visualize_classification_groups.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+		selected_inds = np.random.choice(inds, num_imgs, replace=False)				# choose a subset of max 30 elements from this set
+		print(f'Number of images in {slidename}: {len(inds)}')
+		blank_slide_layout = ppt.slide_layouts[6]
+		slide = ppt.slides.add_slide(blank_slide_layout)
+		txBox = slide.shapes.add_textbox(Inches(ppt.slide_height.inches/2), Inches(0.2), Inches(2), Inches(1))	# add slide title
+		txBox.text_frame.text = slidename
+		for j, ind in tqdm(enumerate(selected_inds)):
+			coord = cartesian_coords[j]
+			imgnew, label, imgold = dset[ind]
+			with tables.open_file(dset.fname, 'r') as f:	# we need to get the image_filename manually.
```
Comment (choosehappy) [inline visualize_classification_groups.py]: here you're opening the table, but the table is already open in the background because of the dataloader. while for reading this shouldn't cause any problems (Though i have seen it with hdf5) this is considered bad practice as there may be some race conditions if the underlying library isn't threadsafe

File: visualize_classification_groups.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+	bn_size=4
+	drop_rate=0
+
+	##### COMPOSE IMAGE TRANSFORM OBJECT #####
+	img_transform = Compose([	# we choose not to apply any image transforms.
+		# VerticalFlip(p=.5),
+		# HorizontalFlip(p=.5),
+		# HueSaturationValue(hue_shift_limit=(-25,0),sat_shift_limit=0,val_shift_limit=0,p=1),
+		# Rotate(p=1, border_mode=cv2.BORDER_CONSTANT,value=0),
+		RandomSizedCrop((args.patch_size,args.patch_size), args.patch_size,args.patch_size),
```
Comment (choosehappy) [inline visualize_classification_groups.py]: seems like a strange choice, isn't the randomsized crop going to randomly resize and crop? did you look at the documentation? the range seems to be pretty large, from .08 to 1?

https://albumentations.ai/docs/api_reference/augmentations/crops/transforms/#albumentations.augmentations.crops.transforms.RandomResizedCrop

perhaps you want something more like CenterCrop or with padding?

**Review by choosehappy — COMMENTED** [review]
I'm wondering, can we make our lives a bit easier here and split this into two scripts to improve overall clarity and usability by 3rd parties?

Script 1  -  compute and store 'predict' column - this has all the DL stuff
Script 2 - make the pptx with the assumption that the predict column is already in place - this has no DL components and thus we don't need pytorch, albumentations, etc etc making it much lighter

Looking at the code you have, it seems like you should be able to do this in a few minutes with some cut->pasting

**Comment by jacksonjacobs1** [comment]
Here are the changes I've made:

- split the original script into 2 separate scripts:
Script 1: generate_densenet_predictions.py takes a pytable and a densenet checkpoint and generates a "predictions" dataset in the pytable.
Script 2: visualize_classification_groups.py takes a pytable and a powerpoint path, and generates the powerpoint.
- replaced cartesian_coordinates() with np.mgrid[] to compute a coordinates grid
- changed required CLI arguments to positional
- image transform now uses center crop instead of random crop.

File: generate_densenet_predictions.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+	for img, _, _ in tqdm(dloader):
+		img = img.to(device)
+		pred = model(img)
+		p=pred.detach().cpu().numpy()
+		predflat=np.argmax(p,axis=1).flatten()
+		predictions.append(predflat)
+		
+	# save predictions to h5
+	with tables.open_file(args.pytable_path, 'a') as f:
+		try:
```
Comment (choosehappy) [inline generate_densenet_predictions.py]: this isn't the ideal way of doing this type of process and could cause you pain later on if e.g., create_carray deletes what is already present and you don't want it to

essentially the correct software pattern is "check to see if exists" as opposed to catching an error if creating it fails

another point is that try/except patterns require a lot of additional overhead and as such the "except" clause should only be rarely executed.

this is the difference between proactive programming and 'reactive'

so here you want to do something like:

if not table["predictions"]: --- > make table

this also has no "else" associated with it making it a bit cleaner to understand logically

File: visualize_classification_groups.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+	parser.add_argument('pytable_path', type=str, help='The path to an h5 file that contains "patch", "ground_truth_label", and "fname" datasets.')	
+	parser.add_argument('ppt_save', type=str, help='The full path to save the generated powerpoint.')
+
+	parser.add_argument('--str_criteria', '-s', default=["TRUE NEGATIVES", "FALSE NEGATIVES", "FALSE POSITIVES", "TRUE POSITIVES"], type=str, nargs="+")
+	parser.add_argument('--criteria', default=['00', '10', '01', '11'], type=str, nargs="+", help="Each item is a pair of ground-truth,prediction labels.")
+	args = parser.parse_args()
+	
+
+	##### INITIALIZE CRITERIA DICT #####
+	criteria_dict = {}
```
Comment (choosehappy) [inline visualize_classification_groups.py]: you can actually do this entire section in 1 line : ) 

```
criteria_dict = {args.str_criteria[i]: [int(j) for j in args.criteria[i]] for i in range(len(args.str_criteria))}
```

is it more readable? perhaps not, but this works: )

**Review by choosehappy — COMMENTED** [review]
YEA! two minor things and then i think we're all set!
